### ISSUE TITLE
Socket (lowest_layer) close does not cancel SSL pending operations

### ISSUE THREAD

**Issue by kiwixz** [chriskohlhoff/asio#355]
`cancel` does not work either, neither do an incomplete `async_shutdown` on the SSL stream.
I want to close it ungracefully (without waiting for ACK).

Is it expected behavior ?  If so, is there a way to cancel/close SSL stream ?

**Comment by dmxvlx** [comment]
> Is it expected behavior ?
Yes, if socket/stream's dtor was called before async_shutdown, I guess...

> If so, is there a way to cancel/close SSL stream ?
Regular 'lowest_layer().shutdown()' works for me fine.

--
the best regards

**Comment by kiwixz** [comment]
Unfortunately `lowest_layer().shutdown(both)` is not enough for me.

Attempting to delete SSL stream just after `shutdown`/`close` (if you wait a little it "works") leads to crash because apparently some recv is still going on...

I expected no callback to still be going on after a `cancel`/`close` because the documentation states:
>  Any asynchronous send, receive or connect operations will be cancelled **immediately**, and will complete with the asio::error::operation_aborted error.

However it does not explicitly specify about stream operations.

**Comment by dmxvlx** [comment]
I have classes: session_ssl, server_ssl, client_ssl.

In the class session_ssl defined async handlers for read/write
with enabled socket's option nonblocking + queuing of packets.
All works fine with shutdown_both() + close/cancel in ~session_ssl.

I do not know what wrong in your code - I am not seeing that.
I guess there are mixing sync/async operations, problems with multithreading,
or something other.

> leads to crash because apparently some recv is still going on...

Perhaps you need to stop io_context, try it (after shutdown+close):
if (is_client_ && !socket().get_io_context().stopped())
socket().get_io_context().stop();

--
the best regards

**Comment by chriku** [comment]
This issue persists even in (boost version) 1.81.0:
When cancelling the lowest_layer, the handler is only called after calling either `io_context.poll` or waiting/`post`ing once.
This happens a single_threaded `io_context` and is especially horrible if it is caused by calling the destructor by exceptions through coroutines, as the handler then uses a deleted `ssl::stream` and segfaults into the afterlife.

waiting for the handler (by `lowest_layer().cancel()` or `shutdown`) is not practical during exception cleanup.
